### PR TITLE
VPN-5466 - Bring back LAN Bypass setting on iOS

### PR DESCRIPTION
**Pull request by brizental**
This is here mostly for illustration purposes for now, therefore it's a draft and doesn't really require a review.

Notes:

1. CheckBox enabled setting is not doing anything. @MattLichtenstein anything that immediatelly pops out? TBH, I did not deeply investigate it given the state of this PR, but if already know what it is, good.
2. Do we want to add telemetry to the new button?
3. Can't write tests for this, because it's iOS and we don't run tests there yet.
4. A lot of copy paste from https://github.com/mozilla-mobile/mozilla-vpn-client/pull/5096

Screenshot: 
Screenshot 2024-02-02 at 13 05 19

### CODE REVIEW

**Comment by MattLichtenstein** [comment]
> CheckBox enabled setting is not doing anything. @MattLichtenstein anything that immediatelly pops out? TBH, I did not deeply investigate it given the state of this PR, but if already know what it is, good.

The issue is you have the `enabled` property bound to an enum instead of an expression ([here](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/9070/files#diff-acab1d9449771c3371d8f4e7f6fc0309b9a9931e1077913d647a85dae31b7f01R87))

actual: `enabled: VPNController.StateOff`

expected: `enabled: VPNController.state === VPNController.StateOff`

You can also get rid of the `enabled` binding that you added to the MZCheckBox [here](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/9070/files#diff-513b45286f8b5b56dc463b7f7bc4d166e3ab5c0cf9bd92e2006f27a0c71c3c55R33) since disabling the MZCheckBoxRow disables all of it's children

**Comment by brizental** [comment]
> The issue is you have the enabled property bound to an enum instead of an expression ([here](https://github.com/mozilla-mobile/mozilla-vpn-client/pull/9070/files#diff-acab1d9449771c3371d8f4e7f6fc0309b9a9931e1077913d647a85dae31b7f01R87))

:woman_facepalming:

**Comment by brizental** [comment]
There are two commits I sent and are not showing up here yet @lesleyjanenorton , maybe wait a bit and if there are not two commits after https://github.com/mozilla-mobile/mozilla-vpn-client/pull/9070/commits/5b3fa0ec5ce29ad0e3109475a9b01bce21964403 then this is not up to date and I'll see what is up tomorrow 🤦‍♀️

File: src/settingslist.h [hunk 1]
```diff
@@ -709,6 +709,18 @@ SETTING_STRINGLIST(subscriptionTransactions,        // getter
                    false,                           // remove when reset
                    true                             // sensitive (do not log)
 )
+
+SETTING_BOOL(localNetworkAccess,        // getter
+             setLocalNetworkAccess,     // setter
+             removeLocalNetworkAccess,  // remover
+             hasLocalNetworkAccess,     // has
+             "localNetworkAccess",      // key
+             true,                      // default value
```
Comment (brizental) [inline src/settingslist.h:718]: TODO: Change this to `false`.

File: src/translations/strings.yaml [hunk 2]
```diff
@@ -360,6 +360,10 @@ settings:
     value: Clear all
     comment: Button label to clear all apps from the exclusions list
 
+localNetworkAccess:
+  label: "Local network access"
```
Comment (flodolo) [inline src/translations/strings.yaml:364]: Why the quotes?

File: src/translations/strings.yaml [hunk 3]
```diff
@@ -360,6 +360,10 @@ settings:
...
+  subLabel: "Access printers, streaming sticks and all other devices on your local network"
```
[the remark this answers is not in the document]
Comment (flodolo) [reply src/translations/strings.yaml]: No, same IDs is fine.

File: src/translations/strings.yaml (hunk 2 above)
Comment (brizental) [reply src/translations/strings.yaml:364]: Copy/Paste :spaghetti: Will remove.

File: src/ui/screens/settings/ViewPreferences.qml [hunk 4]
```diff
@@ -67,6 +68,27 @@ MZViewBase {
...
+                anchors {
+                    right: parent.right
+                    left: parent.left
+                    rightMargin: VPNTheme.theme.windowMargin
+                }
+
+                labelText: MZI18n.LocalNetworkAccessLabel
+                subLabelText: MZI18n.LocalNetworkAccessSubLabel
+                isChecked: MZSettings.localNetworkAccess
+                enabled: VPNController.state === VPNController.StateOff
```
Comment (mcleinman) [inline src/ui/screens/settings/ViewPreferences.qml:84]: Should part of the text explain that this is only enabled when the VPN is off? If not, users will show up here with the VPN active and have no indication as to why the setting is disabled.

**Review by mcleinman — CHANGES_REQUESTED** [review]
This looks great overall, just a couple questions.

File: src/ui/screens/settings/ViewPreferences.qml (hunk 4 above)
Comment (brizental) [reply src/ui/screens/settings/ViewPreferences.qml:84]: That is a good point. There are other features that depend on the VPN being turned off. I'll check how they handle it.

Comment (brizental) [reply src/ui/screens/settings/ViewPreferences.qml:84]: Interesting, we actually don't have this behaviour anywhere in the app. I am pretty sure we used to. I'll ask today during stand up.

File: src/ui/screens/settings/ViewPreferences.qml [hunk 5]
```diff
@@ -17,22 +18,25 @@ MZViewBase {
     property string _languageTitle: ""
     readonly property string telemetryScreenId : "app_preferences"
 
+    property bool vpnIsOff: VPNController.state === VPNController.StateOff
+    property bool isIOS: Qt.platform.os === "ios"
```
Comment (mcleinman) [inline src/ui/screens/settings/ViewPreferences.qml:22]: Nit: Move these two to `MZUiUtils.js`?

File: src/platforms/ios/ioscontroller.swift [hunk 6]
```diff
@@ -116,7 +116,7 @@ public class IOSControllerImpl : NSObject {
         }
     }
 
-    @objc func connect(dnsServer: String, serverIpv6Gateway: String, serverPublicKey: String, serverIpv4AddrIn: String, serverPort: Int,  allowedIPAddressRanges: Array<VPNIPAddressRange>, reason: Int, gleanDebugTag: String, isSuperDooperFeatureActive: Bool, installationId: String, disconnectOnErrorCallback: @escaping () -> Void, onboardingCompletedCallback: @escaping () -> Void, vpnConfigPermissionResponseCallback: @escaping (Bool) -> Void) {
+    @objc func connect(dnsServer: String, serverIpv6Gateway: String, serverPublicKey: String, serverIpv4AddrIn: String, serverPort: Int,  excludeLocalNetworks: Bool, allowedIPAddressRanges: Array<VPNIPAddressRange>, reason: Int, gleanDebugTag: String, isSuperDooperFeatureActive: Bool, installationId: String, disconnectOnErrorCallback: @escaping () -> Void, onboardingCompletedCallback: @escaping () -> Void, vpnConfigPermissionResponseCallback: @escaping (Bool) -> Void) {
```
Comment (mcleinman) [inline src/platforms/ios/ioscontroller.swift:119]:
```suggestion
    @objc func connect(dnsServer: String, serverIpv6Gateway: String, serverPublicKey: String, serverIpv4AddrIn: String, serverPort: Int,  allowedIPAddressRanges: Array<VPNIPAddressRange>, reason: Int, excludeLocalNetworks: Bool, gleanDebugTag: String, isSuperDooperFeatureActive: Bool, installationId: String, disconnectOnErrorCallback: @escaping () -> Void, onboardingCompletedCallback: @escaping () -> Void, vpnConfigPermissionResponseCallback: @escaping (Bool) -> Void) {
```

I believe this fixes the crash

**Review by mcleinman — CHANGES_REQUESTED** [review]
This is crashing every time I try to activate the VPN. I believe this is because the function's arguments are not in the expected order.

**Comment by brizental** [comment]
> This is crashing every time I try to activate the VPN. I believe this is because the function's arguments are not in the expected order.

That is what I get for trusting month old code 🤦‍♀️ Thank you so much for testing this.

**Review by mcleinman — APPROVED** [review]
Looks good, thanks.

**Comment by ValentinaPC** [comment]
@brizental Does this have a Jira ticket? Thanks!

**Pull request merged**